edition = "2023";
package circuitsolver;

message CircuitGraphMessage {
  repeated Edge edges = 1;
  repeated Vertex vertices = 2;

  message Edge {
    int32 id = 1;
    int32 fromId = 2;
    int32 toId = 3;
    double current = 4; // Optional
    oneof specificBranch {
      CurrentSource currentSource = 5;
      IdealDiode idealDiode = 6;
      RealDiode realDiode = 7;
      Resistor resistor = 8;
      VoltageSource voltageSource = 9;
      ZenerDiode zenerDiode = 10;
    }

  message CurrentSource {
    double voltage = 1; // Optional
  }
  message IdealDiode {
    double voltage = 1; // Optional
  }
  message RealDiode {
    double i0 = 1; // Optional
    double vt = 2; // Optional
    double n = 3; // Optional
  }
  message Resistor {
    double resistance = 1; // Optional
  }
  message VoltageSource {
    double voltage = 1; // Optional
  }
  message ZenerDiode {
    double vzt = 1; // Optional
    double rzt = 2; // Optional
    double izt = 3; // Optional
  }

  }
  message Vertex {
    int32 id = 1;
    double voltage = 2; // Optional
  }
}

